src/rust/engine/src/externs.rs

```diff
@@ -462,23 +452,18 @@ impl From<Result<(), String>> for PyResult {
   }
 }
 
-// Only constructed from the python side.
-// TODO: map this into a C enum with cbindgen and consume from python instead of using magic numbers
-// in extern_generator_send() in native.py!
-#[allow(dead_code)]
-#[repr(u8)]
-pub enum PyGeneratorResponseType {
-  Break = 0,
-  Throw = 1,
-  Get = 2,
-  GetMulti = 3,
-}
-
+///
+/// The response from a call to extern_generator_send. Gets include Idents for their Handles
+/// in order to avoid roundtripping to intern them, and to eagerly trigger errors for unhashable
+/// types on the python side where possible.
+///
 #[repr(C)]
-pub struct PyGeneratorResponse {
-  res_type: PyGeneratorResponseType,
-  values: HandleBuffer,
-  products: TypeIdBuffer,
+pub enum PyGeneratorResponse {
+  Get(TypeId, Handle, Ident),
+  GetMulti(TypeIdBuffer, HandleBuffer, IdentBuffer),
+  // NB: Broke not Break because C keyword.
+  Broke(Handle),
+  Throw(Handle),
```

Comment:
Much better!! Thank you for taking the time to figure out how to do this with cffi, it makes the code leagues more readable. It also obviates the need for that TODO without additional complexity in building the engine!